crates/matrix-sdk/src/client/mod.rs

```diff
@@ -437,48 +439,77 @@ impl Client {
-                        }
-                        Ok(None) => {
-                            error!(
-                                "Event handler for {} has an invalid context argument",
-                                event_type
-                            );
-                        }
-                        Err(e) => {
-                            warn!(
-                                "Failed to deserialize `{}` event, skipping event handler.\n\
-                                 Deserialization error: {}",
-                                event_type, e,
-                            );
+
+        let handle = EventHandlerHandle::new(H::ID);
+
+        self.inner.event_handlers.write().await.entry(H::ID).or_default().push(
+            EventHandlerWrapper {
+                handler_function: Box::new(move |data| {
+                    let maybe_fut = serde_json::from_str(data.raw.get())
+                        .map(|ev| handler.clone().handle_event(ev, data));
+
+                    Box::pin(async move {
+                        match maybe_fut {
+                            Ok(Some(fut)) => {
+                                fut.await.print_error(event_type);
+                            }
+                            Ok(None) => {
+                                error!(
+                                    "Event handler for {} has an invalid context argument",
+                                    event_type
+                                );
+                            }
+                            Err(e) => {
+                                warn!(
+                                    "Failed to deserialize `{}` event, skipping event handler.\n\
+                                        Deserialization error: {}",
+                                    event_type, e,
+                                );
+                            }
                         }
-                    }
-                })
+                    })
+                }),
+                handle: handle.clone(),
             },
-        ));
+        );
 
-        self
+        handle
     }
 
     pub(crate) async fn event_handlers(&self) -> RwLockReadGuard<'_, EventHandlerMap> {
         self.inner.event_handlers.read().await
     }
 
+    /// Remove the event handler associated with the token.
+    ///
+    ///  # Arguments
+    ///
+    /// `handle` - The [`EventHandlerHandle`] that was passed to
+    /// [`Client::register_event_handler`] when registering the handler.
+    pub async fn remove_event_handler(&self, handle: EventHandlerHandle) -> &Self {
+        let event_id = handle.1;
+        let mut empty = false;
+
+        if let Some(v) = self.inner.event_handlers.write().await.get_mut(&event_id) {
+            v.retain(|e| -> bool { e.handle.0.ne(&handle.0) });
+
+            empty = v.is_empty();
+        }
+
+        if empty {
+            self.inner.event_handlers.write().await.remove(&event_id);
+        }
```

Comment:
Please only lock the event handler `RwLock` once here. Also, instead of the separate `empty` variable, you can just have a `if v.is_empty()` branch inside the `if let`.